src/main/scala/org/kframework/kore/interfaces/Interface.scala

```diff
@@ -0,0 +1,303 @@
+  }
+
+  trait Or extends Pattern {
+    val p: Pattern
+    val q: Pattern
+  }
+
+  object Or {
+    def unapply(arg: Or): Option[(Pattern, Pattern)] = Some(arg.p, arg.q)
+  }
+
+  trait Not extends Pattern {
+    def p: Pattern
+  }
+
+  object Not {
+    def unapply(arg: Not): Option[Pattern] = Some(arg.p)
+  }
+
+  trait Implies extends Pattern {
+    val p: Pattern
+    val q: Pattern
+  }
+
+  object Implies {
+    def unapply(arg: Implies): Option[(Pattern, Pattern)] = Some(arg.p, arg.q)
+  }
+
+  trait Exists extends Pattern {
+    val v: Variable
+    val p: Pattern
+  }
+
+  object Exists {
+    def unapply(arg: Exists): Option[(Variable, Pattern)] = Some(arg.v, arg.p)
+  }
+
+  trait ForAll extends Pattern {
+    val v: Variable
+    val p: Pattern
+  }
+
+  object ForAll {
+    def unapply(arg: ForAll): Option[(Variable, Pattern)] = Some(arg.v, arg.p)
+  }
+
+  trait Next extends Pattern {
+    def p: Pattern
+  }
+
+  object Next {
+    def unapply(arg: Next): Option[Pattern] = Some(arg.p)
+  }
+
+
+  trait Rewrite extends Pattern {
+    val p: Pattern
+    val q: Pattern
+  }
+
+  object Rewrite {
+    def unapply(arg: Rewrite): Option[(Pattern, Pattern)] = Some(arg.p, arg.q)
+  }
+
+  trait Equals extends Pattern {
+    val p: Pattern
+    val q: Pattern
+  }
+
+  object Equals {
+    def unapply(arg: Equals): Option[(Pattern, Pattern)] = Some(arg.p, arg.q)
+  }
+
+  trait ModuleName {
+    def str: String
+  }
+
+  object ModuleName {
+    def unapply(arg: ModuleName): Option[String] = Some(arg.str)
+  }
+
+  trait Name {
+    val str: String
+  }
+
+  object Name {
+    def unapply(arg: Name): Option[String] = Some(arg.str)
+  }
+
+  trait Sort {
+    val str: String
+  }
+
+  object Sort {
+    def unapply(arg: Sort): Option[String] = Some(arg.str)
+  }
+
+  trait Symbol {
+    val str: String
+  }
+
+  object Symbol {
+    def unapply(arg: Symbol): Option[String] = Some(arg.str)
+  }
+
+  trait Value {
+    val str: String
+  }
+
+  object Value {
+    def unapply(arg: Value): Option[String] = Some(arg.str)
+  }
+
+}
+
+
+trait Builders {
+
+  import Kore._
+
+  def Definition(modules: Seq[Module], att: Attributes): Definition
+
+  def Module(name: ModuleName, sentences: Seq[Sentence], att: Attributes): Module
+
+  def Import(name: ModuleName, att: Attributes): Import
+
+  def SortDeclaration(sort: Sort, att: Attributes): SortDeclaration
+
+  def SymbolDeclaration(sort: Sort, symbol: Symbol, args: Seq[Sort], att: Attributes): SymbolDeclaration
+
+  def Rule(p: Pattern, att: Attributes): Rule
+
+  def Axiom(p: Pattern, att: Attributes): Axiom
+
+  def Attributes(att: Seq[Pattern]): Attributes
+
+  def Variable(name: Name, sort: Sort): Variable
+
+  def Application(p: Symbol, args: Seq[Pattern]): Application
+
+  def DomainValue(symbol: Symbol, value: Value): DomainValue
+
+  def Top(): Top
+
+  def Bottom(): Bottom
+
+  def And(p: Pattern, q: Pattern): And
```

Comment:
Ok, I think we can do what @cos says, and at second thoughts it seems to be the right thing to do.  But, we need to change the syntax of KORE back to:
```
...
Pattern =  Variable
         | \and(Pattern,Pattern)
...
Variable = Name:Sort
```
Here is "meaning" of the translation to Scala:

1) Each production `N = prod(N1,...,Nk)` yields an extractor trait extending the result sort and a companion object for matching purposes:
```
trait Prod extends N {
  val _1 : N1
  ...
  val _k : Nk
}
object Prod {
    def unapply(arg: Prod): Option[(N1,...,Nk)] = Some(arg._1, ..., arg._k)
}
```
as well as a builder:
```
  def Prod(n1: N1, ..., nk: Nk): N
```

2) Each subsort production `N = M` yields a trait extension relation:
```
trait M extends N
```

With these, 
```
Pattern = \and(Pattern,Pattern)
```
yields:
```
  trait And extends Pattern {
    val p: Pattern
    val q: Pattern
  }
  object And {
    def unapply(arg: And): Option[(Pattern, Pattern)] = Some(arg.p, arg.q)
  }
...
def And(p: Pattern, q: Pattern): Pattern
```
Similarly, 
```
  Pattern = Variable
  Variable = Name:Sort
```
yields:
```
  trait Variable extends Pattern
  trait SortedVariable extends Variable {
    val name: Name
    val sort: Sort
  }
  object SortedVariable {
    def unapply(arg: Variable): Option[(Name, Sort)] = Some(arg.name, arg.sort)
  }
...
def SortedVariable(n: Name, s: Sort): Variable  // Variable, not Pattern!
```
One could now argue that we can merge `SortedVariable` with `Variable` because there is only one constructor for `Variable`, but if we had another one, say `UnsortedVariable`, due to some extended KORE syntax `Variable = var(Name)`, then we may want to keep a separation between `Variable` and the two constructs for them.